tests/codex/testvalidation.nim

```diff
@@ -41,12 +56,170 @@ asyncchecksuite "validation":
   test "the list of slots that it's monitoring is empty initially":
     check validation.slots.len == 0
 
+  for (validationGroups, groupIndex) in [(100, 100'u16), (100, 101'u16)]:
+    test "initializing ValidationConfig fails when groupIndex is " &
+        "greater than or equal to validationGroups " &
+        fmt"(testing for {groupIndex = }, {validationGroups = })":
+      let groups = ValidationGroups(validationGroups).some
+      let validationConfig = ValidationConfig.init(
+          maxSlots, groups = groups, groupIndex = groupIndex)
+      check validationConfig.isFailure == true
+      check validationConfig.error.msg == "The value of the group index " &
+          "must be less than validation groups! " &
+          fmt"(got: {groupIndex = }, groups = {!groups})"
+  
+  test "initializing ValidationConfig fails when maxSlots is negative":
+    let maxSlots = -1
+    let validationConfig = ValidationConfig.init(
+        maxSlots = maxSlots, groups = ValidationGroups.none)
+    check validationConfig.isFailure == true
+    check validationConfig.error.msg == "The value of maxSlots must " &
+        fmt"be greater than or equal to 0! (got: {maxSlots})"
+  
+  test "initializing ValidationConfig fails when maxSlots is negative " &
+      "(validationGroups set)":
+    let maxSlots = -1
+    let validationConfig = ValidationConfig.init(
+        maxSlots = maxSlots, groups = validationGroups, groupIndex)
+    check validationConfig.isFailure == true
+    check validationConfig.error.msg == "The value of maxSlots must " &
+        fmt"be greater than or equal to 0! (got: {maxSlots})"
```

Comment:
🟡 This test doesn't cover any additional lines, does it? Maybe it's not needed then?

Reply:
The test before checks for negative maxSlots...

Reply:
Sorry, I put it wrong. I wanted to communicate that `maxSlots` are validated independently from `validationGroups`. But it is also true that it does not cover any additional lines, so if you both say it is not needed I will remove it. The less code the better...

Reply:
Unless there is a case where `validationGroups` is dependant on `maxSlots`, which as you mentioned is validated independently, so I would say please remove it.

Reply:
Indeed `validationGroups` **should** not depend on `maxSlots`. But we should test that. If we only have a test with negative `maxSlots` when `validationGroups` is not set, then it could still happen that there is a bug in the implementation causing the negative value of `maxSlots` to be accepted in the case when `validationGroups` is set. Saying that it is "validated independently" is relaying on the current implementation. If I change the implementation in such a way that negative `maxSlots` is accepted when `validationGroups` is set (which would be a bug), I would prefer to have a failing test.

For instance, if I remove this test, and I will make the implementation of the `ValidationConfig` initialiser the following:

```nim
func init*(
    _: type ValidationConfig,
    maxSlots: MaxSlots,
    groups: ?ValidationGroups,
    groupIndex: uint16 = 0): ?!ValidationConfig =

  if validationGroups =? groups and groupIndex >= uint16(validationGroups):
    return failure "The value of the group index must be less than " &
        fmt"validation groups! (got: {groupIndex = }, " &
        fmt"groups = {validationGroups})"
  
  if groups.isSome and maxSlots < 0:
    return success ValidationConfig(
      maxSlots: maxSlots, groups: groups, groupIndex: groupIndex)

  if maxSlots < 0:
    return failure "The value of maxSlots must be greater than " &
        fmt"or equal to 0! (got: {maxSlots})"
  
  success ValidationConfig(
      maxSlots: maxSlots, groups: groups, groupIndex: groupIndex)
```

then I would have all the tests pass even thought there is an obvious bug.

Reply:
Decided to keep that one for now.